book/src/pil/types.md

```diff
@@ -0,0 +1,163 @@
+chosen when powdr is run. The modulus of that field can be accessed via `std::field::modulus()`.
+
+Field elements are the values stored in (fixed, witness and intermeditae) columns.
+Arithmetic inside identities (algebraic expressions) is also always finite field arithmetic.
+
+Since literal numbers in the source code are integers, those have to be converted to field elements
+using the `std::conv::fe` function. In some situations (see below), integers
+(including literal numbers) are converted to field elements implicitly.
+
+Field elements allow the following operators, where the result is always a field element:
+
+- `+`: finite field addition
+- `-`: finite field subtraction
+- `*`: finite field multiplication
+
+There is also an exponentiation operator on field elements. It requires the exponent
+to be a non-negative integer and thus has the signature `**: fe, int -> fe`.
+
+The following comparison operators exist for field elements, whose result is a boolean:
+
+- `==`: equality comparison
+- `!=`: inequality comparison
+
+Since finite fields do not have an inherent order as integers do, if you want to
+compare them using `<`, you have to first convert them to integers.
+
+### String
+
+Type name: `string`
+
+String literals are written as `"string content"`. They are mainly used for debugging or
+documentation purposes, since they cannot occur in identities.
+
+They allow the following operators:
+
+- `+`: string concatenation
+
+### Tuple
+
+Type name: `(..., ..., ...)`
+
+Tuples are complex types that are composed from other types. The exception is the empty tuple
+type `()`. Examples include `(int, int)` (a pair of integers) and `((fe[], int), ())`
+(a tuple consisting of a tuple that contains an array of field elements and an integer
+and an empty tuple).
+
+Tuples are constructed using parentheses: `(1, 2)` constructs the tuple that consists of
+a one and a two.
+
+Tuples do not allow any operators.
+
+### Array
+
+Type name: `_[]`
+
+Arrays are dynamically-sized collections of elements each of the same type.
+Arrays can be constructed inline using `[1, 2]` (the array containing the
+two elements one and two).
+
+The builtin function `std::array::len` can be used to retrieve the length of an array
+and the elements of an array `a` can be accessed using `a[0]`, `a[1]`, etc.
+
+Arrays allow the following operators:
+
+- `+`: array concatenation
+- `_[]`: array index access. The index needs to be a non-negative integer that is less than the length of the array
+
+
+### Function
+
+Type name: `T1, T2, ..., Tn -> T0`
+
+Functions can be constructed using the lambda expression notation.
+For example `|x, y| x + y` returns a function that performs addition.
+The lambda expression `|| 7` is a function that returns a constant (has no parameters).
+Lambda functions can capture anything in their environment and thus form closures.
+
+Due to the dynamic typing of powdr-pil, the exact type of a lambda expression can often not be determined.
```

Comment:
This sounds confusing. Why does it do dynamic typing?

Reply:
Because nobody found the time to implement a proper generic type checker. This will be replaced once we have a proper type checker. Thibaut told me that I should not document what we plan but what we have...

Reply:
yea but still, we don't do dynamic typing, it's all compile-time

Reply:
Ok, it depends on what you mean by compiling. But even witness generation uses dynamic typing since you can call functions from the query callbacks and those functions could look like
```
|i| match i {
 0 => [],
 1 => 7,
 _ => (1,2)
```

Reply:
wat, really?! Can we stop that :D